File: src/content/algorithms/bucket-sort/code.py
def bucket_sort(arr):  # #@function-def
    if len(arr) == 0:  # #@empty-check
        return arr

    n = len(arr)  # #@buckets-init
    buckets = [[] for _ in range(n)]

    for num in arr:  # #@distribution
        index = min(int(num * n), n - 1)
        buckets[index].append(num)

    for bucket in buckets:  # #@bucket-sort
        bucket.sort()

    result = []  # #@concatenation
    for bucket in buckets:
        result.extend(bucket)

    return result

File: src/content/algorithms/bucket-sort/code.rs
fn bucket_sort(arr: &mut [f64]) {  // #@function-def
    if arr.is_empty() {  // #@empty-check
        return;
    }

    let n = arr.len();  // #@buckets-init
    let mut buckets: Vec<Vec<f64>> = vec![vec![]; n];

    for &num in arr.iter() {  // #@distribution
        let index = ((num * n as f64) as usize).min(n - 1);
        buckets[index].push(num);
    }

    for bucket in &mut buckets {  // #@bucket-sort
        bucket.sort_by(|a, b| a.partial_cmp(b).unwrap());
    }

    let mut idx = 0;  // #@concatenation
    for bucket in buckets {
        for num in bucket {
            arr[idx] = num;
            idx += 1;
        }
    }
}
